metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -953,6 +971,48 @@ private void cancelMaybeFenceReplicas() {
         queue.cancelDeferred(MAYBE_FENCE_REPLICAS);
     }
 
+    private static final int MAX_ELECTIONS_PER_IMBALANCE = 1_000;
+
+    private void maybeScheduleNextBalancePartitionLeaders() {
+        final String maybeBalancePartitionLeaders = "maybeBalancePartitionLeaders";
+
+        if (imbalancedScheduled != ImbalanceSchedule.SCHEDULED &&
+            leaderImbalanceCheckIntervalNs >= 0 &&
+            replicationControl.arePartitionLeadersImbalanced()) {
+
+            log.debug(
+                "Scheduling deferred event for {} because scheduled ({}), checkIntervalNs ({}) and isImbalanced ({})",
+                maybeBalancePartitionLeaders,
+                imbalancedScheduled,
+                leaderImbalanceCheckIntervalNs,
+                replicationControl.arePartitionLeadersImbalanced()
+            );
+            long delayNs = time.nanoseconds();
+            if (imbalancedScheduled == ImbalanceSchedule.DEFERRED) {
+                delayNs += leaderImbalanceCheckIntervalNs;
+            }
+            scheduleDeferredWriteEvent(maybeBalancePartitionLeaders, delayNs, () -> {
```

Comment:
Looking down in the event code, I see this comment when we are handling deferred events.

```
                                // The deferred event is ready to run.  Prepend it to the
                                // queue.  (The value for deferred events is a schedule time
                                // rather than a timeout.)

```

In the case of the IMMEDIATE state where we set deadline to the current time, this might mean that it gets prepended to the queue and run before other waiting write events. Is this what we want here? Maybe we should schedule a regular write event for the IMMEDIATE case.

Reply:
Hmm. Yeah, I think that `KafkaEventQueue` is unfair if we schedule deferred events faster than we can process none deferred event. We have a similar problem with snapshot generation. Unfortunately, we can only cancel events that have been deferred.

I think we need to fix this issue in `KafkaEventQueue`.